### PR TITLE
Fixed #16892

### PR DESCRIPTION
**Pull request by elw00d**
Reworked pull request https://github.com/mono/mono/pull/1265

fixes https://bugzilla.xamarin.com/show_bug.cgi?id=16892

### CODE REVIEW

File: mcs/class/Mono.Data.Tds/Mono.Data.Tds.Protocol/Tds70.cs [hunk 1]
```diff
@@ -559,9 +559,8 @@ private void WriteParameterInfo (TdsMetaParameter param)
 			 * If the value is null, not setting the size to 0 will cause varchar
 			 * fields to get inserted as an empty string rather than an null.
 			 */
-			if (colType != TdsColumnType.IntN && colType != TdsColumnType.DateTimeN) {
```
Comment (akoeplinger) [inline mcs/class/Mono.Data.Tds/Mono.Data.Tds.Protocol/Tds70.cs]: why did you remove this check?

Comment (elw00d) [inline mcs/class/Mono.Data.Tds/Mono.Data.Tds.Protocol/Tds70.cs]: Because it did't fix the problem. It only worked with Int and DateTime types. But when Doubles and another types were used the bug was still existing. This check has been appeared after original PR was offered. I think it was incorrect bugfix that fixed only ints and datetimes.
Correct check is written by @egorovmv in his original PR.

Comment (akoeplinger) [inline mcs/class/Mono.Data.Tds/Mono.Data.Tds.Protocol/Tds70.cs]: As far as I can see the if condition checks whether the column type is *not* a nullable Int/DateTime so it should work with the other types?

Comment (elw00d) [inline mcs/class/Mono.Data.Tds/Mono.Data.Tds.Protocol/Tds70.cs]: Code `size=0` should be called only when dealing with char/varchar and other text types (comment `If the value is null, not setting the size to 0 will cause varchar fields to get inserted as an empty string rather than an null.` tells us about it)
So, condition `if (colType != TdsColumnType.IntN && colType != TdsColumnType.DateTimeN)` should except not only ints and datetimes, but all non-text types.
And new condition `if (param.IsTextType && (param.Value == null || param.Value == DBNull.Value))` does it. Only text types will trigger `size=0` code.

Comment (akoeplinger) [inline mcs/class/Mono.Data.Tds/Mono.Data.Tds.Protocol/Tds70.cs]: Ok, makes sense. Could you please update the commit message with that explanation and also mention it's an extension of the https://github.com/elw00d/mono/commit/7ed9d55b181b0199483734102dc203a1e629e46b fix.

Comment (elw00d) [inline mcs/class/Mono.Data.Tds/Mono.Data.Tds.Protocol/Tds70.cs]: Done

**Comment by elw00d** [comment]
any thoughts on this ?

**Comment by migueldeicaza** [comment]
Do you have a way of testing if the comment from this bug inserts the wrong data on SQL server?

https://bugzilla.xamarin.com/show_bug.cgi?id=16892#c21

(That is the sample in comment 21)
